sklearn/tree/_criterion.pyx

```diff
@@ -51,11 +51,14 @@ cdef class Criterion:
     def __setstate__(self, d):
         pass
 
-    cdef void init(self, DOUBLE_t* y, SIZE_t y_stride, DOUBLE_t* sample_weight,
-                   double weighted_n_samples, SIZE_t* samples, SIZE_t start,
-                   SIZE_t end) nogil:
+    cdef int init(self, DOUBLE_t* y, SIZE_t y_stride, DOUBLE_t* sample_weight,
+                  double weighted_n_samples, SIZE_t* samples, SIZE_t start,
+                  SIZE_t end) nogil except -1:
         """Placeholder for a method which will initialize the criterion.
 
+        Returns -1 in case of failure to alocate memory (and raise MemoryError)
```

Comment:
Looks like this error is in a few places, I won't mark them all for brevity but make sure to get them all :)